### PR TITLE
Hook: Jira plugin does not update PR links when the title changes

### PR DESCRIPTION
**Pull request by danilo-gemoli**
**Abstract**
Once the link between the PR and its related Jira card is set for the first time, the plugin is unable to update it if something like the title changes.

**Description**
This is what happens: the Jira issue did not reference the PR until the first retitle, then I issued the command `/retitle` and the CI component reacted by retitling the PR and linking the it to the issue I referenced (as part of the new title). Unfortunately the steps are not consequential, so I ended up finding the old title in the issue. From that moment on any new `/retitle` would have not affected the link between the Jira issue and the PR because the system is not able to update an already set link.

### CODE REVIEW

**Comment by alvaroaleman** [comment]
/label tide/merge-method-squash

I don't think we ever update the title in jira. Why do we have to delete the link to do that?

**Comment by alvaroaleman** [comment]
From the docs it looks to me like an existing link can be updated: https://docs.atlassian.com/software/jira/docs/api/REST/7.6.1/#api/2/issue-updateRemoteIssueLink

**Comment by danilo-gemoli** [comment]
You are right, in accord with the documentation you have linked, an existing link can just be updated in place.

File: prow/jira/jira.go [hunk 1]
```diff
@@ -190,6 +191,23 @@ func (jc *client) AddRemoteLink(id string, link *jira.RemoteLink) error {
...
+	if err != nil {
+		return fmt.Errorf("failed to construct request: %w", err)
+	}
+	resp, err := jc.upstream.Do(req, nil)
+	if resp != nil {
+		defer resp.Body.Close()
+	}
+	if err != nil {
+		return fmt.Errorf("failed to update link: %w", JiraError(resp, err))
+	}
```
Comment (alvaroaleman) [inline prow/jira/jira.go:206]: Please also check the returned status code, if e.G. statuscode is 503 but you got a valid response, err will be nil

Comment (danilo-gemoli) [reply prow/jira/jira.go:210]: I have allowed 204 responses only as documentation stated it is the only status code to be considered "valid"

**Comment by danilo-gemoli** [comment]
@alvaroaleman I have just realized there is a race condition in the updating process when both `jira` and `retitle` plugins are enabled. In fact, when commenting a pull request with such `/retitle ${JIRA_ISSUE_ID_HERE}: new-title`, hook receives an `issue_comment` event so it modifies the title of the PR and the comment itself by adding a link to the Jira issue. This causes GitHub to dispatch another two events: `pull_request` (PR has been retitled) and `issue_comment` (comment has been modified) which are handled concurrently. At this point the `jira` plugin is concurrently called twice, so it ends up invoking Jira REST API two times with the same request.
I can spot several possible solutions:

- make `jira` plugin thread safe somehow: maybe the simplest solution here is to **carefully** introduce a Mutex such that no two distinct goroutines can concurrently update the same issue
- revisit the entire flow
- leave it as it is

For sure there are other possible solutions that I am not consider still. What do you think about it?

**Comment by alvaroaleman** [comment]
> leave it as it is

Sounds best to me because
a) Synchronizing this doesn't fit the current architecture
b) Jira doesn't ratelimit us and the retitle event is very likely to be processed after the comment event

File: prow/jira/jira.go [hunk 2]
```diff
@@ -190,6 +191,26 @@ func (jc *client) AddRemoteLink(id string, link *jira.RemoteLink) error {
...
+	internalLinkId := fmt.Sprint(link.ID)
+	req, err := jc.upstream.NewRequest("PUT", "rest/api/2/issue/"+id+"/remotelink/"+internalLinkId, link)
+	if err != nil {
+		return fmt.Errorf("failed to construct request: %w", err)
+	}
+	resp, err := jc.upstream.Do(req, nil)
+	if resp != nil {
+		defer resp.Body.Close()
+	}
+	if err != nil || resp.StatusCode != http.StatusNoContent {
```
Comment (alvaroaleman) [inline prow/jira/jira.go:204]: Can we just do this as:
```
if err != nil {
  return err
}
if resp.StatusCode != 204 {
  return fmt.Errorf("wrong status code")
}
```

That is much easier to follow than first allowing both cases and then subdividing between the two inside the `if ` statement

**Pull request merged**